lib/ConsumerImpl.cc

```diff
@@ -1212,6 +1212,9 @@ void ConsumerImpl::closeAsync(ResultCallback originalCallback) {
 const std::string& ConsumerImpl::getName() const { return consumerStr_; }
 
 void ConsumerImpl::shutdown() {
+    if (ackGroupingTrackerPtr_) {
```

Comment:
We can put it to the `cancelTimers()` method and remove the same logic in the `closeAsync()` method.

Reply:
> In the current code, if consumer creation failed, ConsumerImpl::ackGroupingTrackerPtr_ will not close.

BTW, In your case. It seems that the original design was to close resources in the destructor, but he didn't work.

https://github.com/apache/pulsar-client-cpp/blob/2980bbe9d3ef47eef4a1d468c3898b5a989b4df0/lib/AckGroupingTrackerEnabled.h#L48

Reply:
Get it, thanks for your explain.